### PR TITLE
Add stack kernel

### PR DESCRIPTION
**Pull request by MARD1NO**
自动测试1000轮结果：
image

### CODE REVIEW

File: oneflow/core/functional/impl/array_functor.cpp [hunk 1]
```diff
@@ -483,26 +483,74 @@ class ConcatFunctor {
...
-    CHECK_OR_RETURN(dim >= -(ndims + 1) && dim <= ndims)
-        << "( Dimension out of range, expected to be in range of [" << -(ndims + 1) << ", " << ndims
-        << "], but got " << dim << " )";
-    if (dim < 0) { stack_dim = stack_dim + ndims + 1; }
-    TensorTuple expand_inputs(inputs.size());
-    if (inputs.size() == 1) { return ExpandDims(inputs.at(0), stack_dim); }
-    for (int i = 0; i < inputs.size(); ++i) {
-      expand_inputs[i] = JUST(ExpandDims(inputs.at(i), stack_dim));
+    MutableAttrMap attrs;
+    JUST(attrs.SetAttr<int64_t>("axis", axis));
```
Comment (hjchen2) [inline oneflow/core/functional/impl/array_functor.cpp:510]: axis -> stack_dim

File: oneflow/core/functional/impl/array_functor.cpp [hunk 2]
```diff
@@ -483,26 +483,74 @@ class ConcatFunctor {
...
+    int64_t max_dim_size = 0;
     int64_t stack_dim = dim;
-    for (int i = 1; i < inputs.size(); ++i) {
-      CHECK_EQ_OR_RETURN(inputs.at(i)->shape()->NumAxes(), ndims)
-          << "The input dimensions are not equal.";
+    if (dim < 0) { stack_dim = stack_dim + ndims; }
+    const std::shared_ptr<const Shape>& first_in_shape = inputs[0]->shape();
+    for (const auto& input : inputs) {
+      for (int i = 0; i < ndims; ++i) {
+        if (axis == i) { max_dim_size += 1; }
```
[the remark this answers is not in the document]
Comment (hjchen2) [reply oneflow/core/functional/impl/array_functor.cpp:502]: max_dim_size是什么作用，它就等于输入tensor个数？

File: oneflow/core/autograd/gradient_funcs/stack_op.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+Maybe<void> Stack::Capture(StackCaptureState* ctx, const TensorTuple& inputs,
+                           const TensorTuple& outputs, const AttrMap& attrs) const {
+  ctx->requires_grad.resize(inputs.size());
+  for (int i = 0; i < inputs.size(); ++i) { ctx->requires_grad[i] = inputs.at(i)->requires_grad(); }
+
+  ComposedAttrMap composed_attrs(attrs, base_attrs_);
+  ctx->axis = JUST(composed_attrs.GetAttr<int64_t>("axis"));
+  for (const auto& input : inputs) { ctx->SaveTensorForBackward(input); }
```
Comment (hjchen2) [inline oneflow/core/autograd/gradient_funcs/stack_op.cpp:57]: 这里捕获输入tensor可能造成显存占用变大，感觉我们可能需要两套stack grad算子，一个用于lazy，一个用于eager，他们之间的区别是lazy算子依赖输入tensors，eager算子不依赖输入tensors，但增加一个属性值表示每个tensor的size

File: oneflow/user/kernels/stack_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    ctx->MutShapeView4ArgNameAndIndex("out", 0)->set_shape(Shape(out_dim_vec));
+  }
+
+  void Compute(user_op::KernelComputeContext* ctx) const override {
+    user_op::Tensor* out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    if (out_tensor->shape().elem_cnt() == 0) { return; }
+    const int64_t axis = ctx->Attr<int64_t>("axis");
+    const int64_t out_cols = out_tensor->shape().Count(axis);
+    const int64_t rows = out_tensor->shape().elem_cnt() / out_tensor->shape().Count(axis);
+    CHECK_GT(rows, 0);
```
Comment (hjchen2) [inline oneflow/user/kernels/stack_kernel.cpp:70]: 这里要加一些报错信息

File: oneflow/user/kernels/stack_kernel.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    auto primitive = NewCopyNdPrimitive(ctx);
+    CHECK(primitive);
```
Comment (hjchen2) [inline oneflow/user/kernels/stack_kernel.cpp:72]: 加报错信息

File: oneflow/user/kernels/stack_kernel.cpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    int64_t out_col_offset = 0;
+    for (const auto& in_arg_pair : ctx->inputs()) {
+      const user_op::Tensor* in_tensor =
+          ctx->Tensor4ArgNameAndIndex(in_arg_pair.first, in_arg_pair.second);
+      if (in_tensor->shape().elem_cnt() == 0) { continue; }
+      const int64_t in_cols = in_tensor->shape().Count(axis);
+      CHECK_EQ(in_tensor->shape().elem_cnt(), rows * in_cols);
```
Comment (hjchen2) [inline oneflow/user/kernels/stack_kernel.cpp:79]: 报错信息

File: oneflow/user/kernels/stack_kernel.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        DimVector src_shape = {rows, in_cols};
+        DimVector src_pos_vec = {0, 0};
+        DimVector extent_vec = {rows, in_cols};
+        primitive->Launch(ctx->stream(), out_tensor->data_type(), 2, out_tensor->mut_dptr(),
+                          dst_shape.data(), dst_pos_vec.data(), in_tensor->dptr(), src_shape.data(),
+                          src_pos_vec.data(), extent_vec.data());
+      }
+      out_col_offset += in_cols;
+    }
+    CHECK_EQ(out_col_offset, out_cols);
```
Comment (hjchen2) [inline oneflow/user/kernels/stack_kernel.cpp:92]: 报错信息

File: oneflow/user/kernels/stack_kernel.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  ~StackBackwardKernel() override = default;
+
+ private:
+  void InferShape(user_op::KernelInferContext* ctx) const override {
+    const auto axis = ctx->Attr<int64_t>("axis");
+    const ShapeView& in_shape_view = ctx->ShapeView4ArgNameAndIndex("in", 0);
+    int64_t total_dim_size = 0;
+    const int64_t like_num_axes = ctx->ShapeView4ArgNameAndIndex("like", 0).NumAxes();
+    const int64_t in_num_axes = in_shape_view.NumAxes();
+    CHECK_LE(like_num_axes, in_num_axes);
```
Comment (hjchen2) [inline oneflow/user/kernels/stack_kernel.cpp:122]: 下面这些CHECK都加一下报错信息

File: oneflow/core/autograd/gradient_funcs/stack_op.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+#include "oneflow/core/framework/op_expr.h"
+#include "oneflow/core/framework/op_interpreter/op_interpreter_util.h"
+#include "oneflow/core/functional/functional.h"
+
+namespace oneflow {
+namespace one {
+
+struct StackCaptureState : public AutoGradCaptureState {
+  std::vector<bool> requires_grad;
+  int64_t axis;
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/stack_op.cpp:27]: 默认初始化

File: oneflow/core/autograd/gradient_funcs/stack_op.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  TensorTuple like(ctx->input_num);
+  for (int i = 0; i < ctx->input_num; ++i) { like[i] = ctx->SavedTensors().at(i); }
+  if (ctx->input_num == 1) {
+    in_grads->at(0) = out_grads.at(0);
+  } else {
+    // const auto& results = JUST(functional::SplitLike(out_grads.at(0), like, ctx->axis));
+    const auto& results = JUST(functional::StackBackward(out_grads.at(0), like, ctx->axis));
+    CHECK_EQ_OR_RETURN(results->size(), ctx->input_num);
+
+    for (int i = 0; i < ctx->input_num; ++i)
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/stack_op.cpp:75]: for loop后加{

}

File: oneflow/core/autograd/gradient_funcs/stack_op.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+Maybe<void> Stack::Apply(const StackCaptureState* ctx, const TensorTuple& out_grads,
+                         TensorTuple* in_grads) const {
+  CHECK_EQ_OR_RETURN(out_grads.size(), 1);
+  in_grads->resize(ctx->input_num);
+  TensorTuple like(ctx->input_num);
+  for (int i = 0; i < ctx->input_num; ++i) { like[i] = ctx->SavedTensors().at(i); }
+  if (ctx->input_num == 1) {
+    in_grads->at(0) = out_grads.at(0);
+  } else {
+    // const auto& results = JUST(functional::SplitLike(out_grads.at(0), like, ctx->axis));
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/stack_op.cpp:71]:
```suggestion
    
```

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td [hunk 12]
```diff
@@ -8697,3 +8697,48 @@ def OneFlow_UpsampleTrilinear3DGradOp : OneFlow_BaseOp<"upsample_trilinear_3d_gr
 }
 
 #endif // GET_ONEFLOW_UPSAMPLE_OP_DEFINITIONS
+
+// Group: STACK
```
Comment (BBuf) [inline oneflow/ir/include/OneFlow/OneFlowUserOps.td:8701]: 和Concat放到一个组里。

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td [hunk 13]
```diff
@@ -8697,3 +8697,48 @@ def OneFlow_UpsampleTrilinear3DGradOp : OneFlow_BaseOp<"upsample_trilinear_3d_gr
...
+// stack, stack_backward
+// Total: 2
+
+#ifdef GET_ONEFLOW_STACK_OP_DEFINITIONS
```
Comment (BBuf) [inline oneflow/ir/include/OneFlow/OneFlowUserOps.td:8705]: 这里也不要了。

File: oneflow/user/kernels/stack_kernel.cpp [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    for (const auto& in_arg_pair : ctx->inputs()) {
+      const ShapeView& input_shape_view =
+          ctx->ShapeView4ArgNameAndIndex(in_arg_pair.first, in_arg_pair.second);
+      CHECK_EQ(input_shape_view.NumAxes(), first_input_shape_view.NumAxes());
+      FOR_RANGE(int64_t, i, 0, in_num_axes + 1) {
+        if (i == axis) {
+          out_dim_vec.at(axis) += 1;
+        } else if (i < axis) {
+          CHECK_EQ(input_shape_view.At(i), out_dim_vec.at(i));
```
Comment (BBuf) [inline oneflow/user/kernels/stack_kernel.cpp:54]: 报错信息。

File: oneflow/user/kernels/stack_kernel.cpp [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+
+    ctx->MutShapeView4ArgNameAndIndex("out", 0)->set_shape(Shape(out_dim_vec));
+  }
+
+  void Compute(user_op::KernelComputeContext* ctx) const override {
+    user_op::Tensor* out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    if (out_tensor->shape().elem_cnt() == 0) { return; }
+    const int64_t axis = ctx->Attr<int64_t>("axis");
+    const int64_t out_cols = out_tensor->shape().Count(axis);
+    const int64_t rows = out_tensor->shape().elem_cnt() / out_tensor->shape().Count(axis);
```
Comment (BBuf) [inline oneflow/user/kernels/stack_kernel.cpp:69]: 是否等价于`const int64_t rows = out_tensor->shape().Count(0, axis-1)` ?

File: oneflow/user/kernels/stack_kernel.cpp [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+      }
+    }
+    CHECK_EQ(total_dim_size, in_shape_view.Count(axis));
+  }
+
+  void Compute(user_op::KernelComputeContext* ctx) const override {
+    const user_op::Tensor* in_tensor = ctx->Tensor4ArgNameAndIndex("in", 0);
+    const auto axis = ctx->Attr<int64_t>("axis");
+    const int64_t in_cols = in_tensor->shape().Count(axis);
+    const int64_t rows = in_tensor->shape().elem_cnt() / in_cols;
```
Comment (BBuf) [inline oneflow/user/kernels/stack_kernel.cpp:152]: 同上

File: oneflow/core/autograd/gradient_funcs/stack_op.cpp (hunk 3 above)
Comment (MARD1NO) [reply oneflow/core/autograd/gradient_funcs/stack_op.cpp:57]: 我有想过这个问题，我下个PR来完成吧

File: oneflow/user/kernels/stack_kernel.cpp (hunk 15 above)
Comment (MARD1NO) [reply oneflow/user/kernels/stack_kernel.cpp:69]: 等价于 `const int64_t rows = out_tensor->shape().Count(0, axis)`

File: oneflow/core/functional/impl/array_functor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MARD1NO) [reply oneflow/core/functional/impl/array_functor.cpp]: 等于输入个数，这里的操作我是参考之前的concat算子，之前静态图会接受dynamic shape，所以会一开始遍历所有的input输入，累积max dim size，按照这个最大值分配显存。后面dynamic shape可以允许形状变化，但是必须比这个最大值小。

File: oneflow/user/kernels/stack_kernel.cpp [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    const auto axis = ctx->Attr<int64_t>("axis");
+    const ShapeView& in_shape_view = ctx->ShapeView4ArgNameAndIndex("in", 0);
+    int64_t total_dim_size = 0;
+    const int64_t like_num_axes = ctx->ShapeView4ArgNameAndIndex("like", 0).NumAxes();
+    const int64_t in_num_axes = in_shape_view.NumAxes();
+    CHECK_LE(like_num_axes, in_num_axes)
+        << "The num axes of `like` tensor should be less equal to num axes of `in` tensor. ";
+    CHECK_LE(axis, like_num_axes)
+        << "The axis should be less equal than num axes of `like` tensor. ";
+    FOR_RANGE(int32_t, i, 0, ctx->outputs().size()) {
```
Comment (BBuf) [inline oneflow/user/kernels/stack_kernel.cpp:136]: 136和141这两个for loop的下标类型要不都统一成int64_t？

File: oneflow/user/kernels/stack_kernel.cpp [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+  void Compute(user_op::KernelComputeContext* ctx) const override {
+    const user_op::Tensor* in_tensor = ctx->Tensor4ArgNameAndIndex("in", 0);
+    const int64_t axis = ctx->Attr<int64_t>("axis");
+    const int64_t in_cols = in_tensor->shape().Count(axis);
+    const int64_t rows = in_tensor->shape().Count(0, axis);
+    CHECK_GT(rows, 0) << "The multiplicative from axis 0 to axis " << axis - 1
+                      << " should be greater than 0. ";
+    auto primitive = NewCopyNdPrimitive(ctx);
+    CHECK(primitive) << "Error in Stack Grad kernel NewCopyNdPrimitive. ";
+    ;
```
Comment (BBuf) [inline oneflow/user/kernels/stack_kernel.cpp:182]:
```suggestion
```

**Comment by BBuf** [comment]
再用nsys看一下是否符合预期，我这里没其它问题了。

File: oneflow/user/kernels/stack_kernel.cpp (hunk 17 above)
Comment (MARD1NO) [reply oneflow/user/kernels/stack_kernel.cpp:136]: 136改成size_t遵守stl容器接口，141我们接口返回的是int64_t

File: oneflow/user/kernels/stack_kernel.cpp (hunk 18 above)
Comment (MARD1NO) [reply oneflow/user/kernels/stack_kernel.cpp:182]: 已修改

File: oneflow/user/ops/stack_op.cpp [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+}  // namespace
+
+/* static */ Maybe<void> StackOp::InferLogicalTensorDesc(user_op::InferContext* ctx) {
+  const user_op::TensorDesc& first_in_desc = ctx->InputTensorDesc("in", 0);
+  const int64_t axis = ctx->Attr<int64_t>("axis");
+  CHECK_GE_OR_RETURN(axis, 0) << "The axis should be greater equal than 0.";
+  const int64_t in_num_axes = first_in_desc.shape().NumAxes();
+  CHECK_LE_OR_RETURN(axis, in_num_axes) "The axis should be greater equal than input num axes.";
```
Comment (liufengwei0103) [inline oneflow/user/ops/stack_op.cpp:55]: 这行注释不对

**Review by liufengwei0103 — APPROVED** [review]

**Comment by MARD1NO** [comment]
> 再用nsys看一下是否符合预期，我这里没其它问题了。

profile耗时正常

**Review by BBuf — APPROVED** [review]
